src/graph_ir.rs

```diff
@@ -211,75 +232,57 @@ impl GraphIR {
         let mut output_buffers = vec![];
 
         // Collect the inputs to the algorithm, the incoming/outgoing edges of this node.
-        let edges = &self.adjacent[&node.id];
-        let inputs = node.inputs.iter().map(|p| (p, true));
-        let outputs = node.outputs.iter().map(|p| (p, false));
+        let adjacent_edges = &self.adjacent[&node.id];
```

Comment:
Just for my understanding the change here is the loop is manually unrolled?

Reply:
The change here is that each item in the `adjacent` hash map now contains a separate list for the incoming and outgoing edges (`NodeEdges`).

Every place this map is used in the algorithm, it is only iterating over the incoming edges or the outgoing edges, so this is more optimal. Plus this makes it nicer for my helper struct.